nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/QueryDatabaseTable.java

```diff
@@ -0,0 +1,608 @@
+
+            fileToProcess = session.create();
+            fileToProcess = session.write(fileToProcess, new OutputStreamCallback() {
+                @Override
+                public void process(final OutputStream out) throws IOException {
+                    try {
+                        logger.debug("Executing query {}", new Object[]{selectQuery});
+                        final ResultSet resultSet = st.executeQuery(selectQuery);
+                        // Max values will be updated in the state property map by the callback
+                        final MaxValueResultSetRowCollector maxValCollector = new MaxValueResultSetRowCollector(statePropertyMap, preProcessStrategy);
+                        nrOfRows.set(JdbcCommon.convertToAvroStream(resultSet, out, tableName, maxValCollector));
+
+                    } catch (final SQLException e) {
+                        throw new ProcessException("Error during database query or conversion of records to Avro", e);
+                    }
+                }
+            });
+
+            if (nrOfRows.get() > 0) {
+                // set attribute how many rows were selected
+                fileToProcess = session.putAttribute(fileToProcess, RESULT_ROW_COUNT, nrOfRows.get().toString());
+
+                logger.info("{} contains {} Avro records; transferring to 'success'",
+                        new Object[]{fileToProcess, nrOfRows.get()});
+                session.getProvenanceReporter().modifyContent(fileToProcess, "Retrieved " + nrOfRows.get() + " rows",
+                        stopWatch.getElapsed(TimeUnit.MILLISECONDS));
+                session.transfer(fileToProcess, REL_SUCCESS);
+            } else {
+                // If there were no rows returned, don't send the flowfile
+                session.remove(fileToProcess);
+                context.yield();
+            }
+
+        } catch (final Throwable /*ProcessException | SQLException*/ e) {
+            if (fileToProcess == null) {
+                // This can happen if any exceptions occur while setting up the connection, statement, etc.
+                logger.error("Unable to execute SQL select query {} due to {}. No FlowFile to route to failure",
+                        new Object[]{selectQuery, e});
+                context.yield();
+            } else {
+
+                logger.error("Unable to execute SQL select query {} due to {}; routing to failure",
+                        new Object[]{selectQuery, e});
+                session.transfer(fileToProcess, REL_FAILURE);
+                context.yield();
+            }
+        } finally {
+            try {
+                session.commit();
+                // Update the state
+                stateManager.setState(statePropertyMap, Scope.CLUSTER);
+            } catch (IOException ioe) {
+                getLogger().error("{} failed to process due to {}; rolling back session", new Object[]{this, ioe});
+                session.rollback(true);
```

Comment:
Not sure that this logic is needed. If setState throws an IOException, calling session.rollback() won't really do anything because session.commit() will have already completed.